src/dialects/db2/query-generator.js

```diff
@@ -751,59 +763,114 @@ export class Db2QueryGenerator extends AbstractQueryGenerator {
-        + 'R.TABNAME AS "tableName", '
-        + 'TRIM(R.TABSCHEMA) AS "tableSchema", LISTAGG(C.COLNAME,\', \') '
-        + 'WITHIN GROUP (ORDER BY C.COLNAME) AS "columnName", '
-        + 'TRIM(R.REFTABSCHEMA) AS "referencedTableSchema", '
-        + 'R.REFTABNAME AS "referencedTableName", '
-        + 'TRIM(R.PK_COLNAMES) AS "referencedColumnName" '
-        + 'FROM SYSCAT.REFERENCES R, SYSCAT.KEYCOLUSE C '
-        + 'WHERE R.CONSTNAME = C.CONSTNAME AND R.TABSCHEMA = C.TABSCHEMA '
-        + `AND R.TABNAME = C.TABNAME${condition} GROUP BY R.REFTABSCHEMA, `
-        + 'R.REFTABNAME, R.TABSCHEMA, R.TABNAME, R.CONSTNAME, R.PK_COLNAMES';
+    const SQL = `
+      SELECT
+              fktable_cat   AS "tableCatalog",
+              fktable_schem AS "tableSchema",
+              fktable_name  AS "tableName",
+              LISTAGG(fkcolumn_name, ', ') WITHIN GROUP (ORDER BY key_seq) AS "tableColumnNames",
+              'FOREIGN KEY' AS "constraintType",
+              fktable_schem AS "constraintSchema",
+              fk_name       AS "constraintName",
+              update_rule   AS "on_update",
+              delete_rule   AS "on_delete",
+              pktable_cat   AS "referencedTableCatalog",
+              pktable_schem AS "referencedTableSchema",
+              pktable_name  AS "referencedTableName",
+              LISTAGG(pkcolumn_name, ', ') WITHIN GROUP (ORDER BY key_seq) AS "referencedTableColumnNames",
+              pk_name       AS "referencedTableConstraintName"
+      FROM    sysibm.sqlforeignkeys
+      ${condition && `WHERE ${condition}`}
+      GROUP BY
+              fktable_cat,
+              fktable_schem,
+              fktable_name,
+              'FOREIGN KEY',
+              fktable_schem,
+              fk_name,
+              update_rule,
+              delete_rule,
+              pktable_cat,
+              pktable_schem,
+              pktable_name,
+              pk_name
+    `;
+
+    return Utils.toSingleLine(SQL);
   }
 
   /**
    * Generates an SQL query that returns all foreign keys of a table.
    *
-   * @param {Stirng|object} table The name of the table.
-   * @param {string} schemaName   The name of the schema.
-   * @returns {string}            The generated sql query.
+   * @param {string|object|Array} table
+   *        The name of the table as a tring
+   *        The tableName and/or schema in an object
+   *        An array of objets
+   *
+   * @param {string} _database
+   *        The name of the database
+   *
+   * @param {object} [options]
+   * @param {string} [options.schema]
+   *        The name of the schema
+   *
+   * @returns {string} The generated sql query.
    */
-  getForeignKeysQuery(table, schemaName) {
-    const tableName = table.tableName || table;
-    schemaName = table.schema || schemaName;
-    let sql = '';
-    if (tableName) {
-      sql = ` AND R.TABNAME = ${wrapSingleQuote(tableName)}`;
-    }
+  getForeignKeysQuery(table, _database, options) {
```

Comment:
this should continue to be `getForeignKeysQuery(table, schemaName)`

The second parameter (used in MariaDB) is the schema name too, no need to pass it through options

Reply:
I think _options_ should always be passed.  I'm came across these changes because CLI migration was not passing along the schema from the config.  I think I had to supply options.

Is there a reason why this couldn't be:
```javascript
getForeignKeysQuery(table, schemaName, options)
```

I'll still have to modify the code to not rely on options, I just don't want to remove it from the equation entirely. We also have some opportunity to fix the taxonomy.

I'm using `table` here because it can be an object or string -- but even that isn't good, I just don't know what's better.  In other places `tableName` is used, which is more specific and I would expect it to be a string, but it can be the object as well.

This is where TypeScript is good because it informs you on the arg types, but even these strings could have schema qualifiers (e.g., `public.users`). So there's multiple equivalent table configurations:
- `users`
- `public.users`
- `{ tableName: 'users' }`
- `{ tableName: 'users', schema: 'users' }`

We see `tableName` is used as a key in the object that points to a string, which is partly why I was confused when I saw it a parameter name that referenced the entire object.

Reply:
I think the issue is that the abstract `QueryInterface.getForeignKeysForTables()` passes databaseName to `getForeignKeysQuery`.  That QI function is a headache.  Maybe each dialect needs its own or instead of passing the databaseName, it should pass the `defaultSchema` created as pat of this commit?

Reply:
> I'm came across these changes because CLI migration was not passing along the schema from the config

This shouldn't prevent the fix from being made. The caller needs to provide the schema name, it shouldn't matter whether it's passed as part of the table name, or as part of option

> Is there a reason why this couldn't be `getForeignKeysQuery(table, schemaName, options)`

Just API design. Our other methods that accept a table name pass the schema as part of the table name (using the `TableName` object). Being able to pass the schema in 3 different ways (as part of the table name, as part of the second parameter, or in options) in very confusing. We should have a clean API, and the other places that call this must make sure they're calling it properly.

The issue with the configured default schema not being passed is separate. We need to normalize the `table` parameter at the beginning of this method. You could do it by creating this new method and calling it with the `table` param:

```typescript
normalizeTableName(tableIdentifier: TableName): TableNameWithSchema {
  if (typeof tableIdentifier === 'string') {
    // normalize to TableNameWithSchema
    tableIdentifier = { tableName: tableIdentifier };
  }

  if (!tableIdentifier.schema) {
    // user provided default value
    if (this.options.schema) {
      tableIdentifier.schema = this.options.schema;
    } else {
       tableIdentifier.schema = this._dialect.getDefaultSchema();
    }
  }

  return tableIdentifier;
}
```

then at the beginning of you method, you can do

```
tableIdentifier = normalizeTableName(tableIdentifier);
```

And it will be much easier to manage, and will have the correct schema name from the config (if set, default schema per dialect otherwise) :)

> I'm using table here because it can be an object or string -- but even that isn't good

`tableName` or `table` are both good. `tableIdentifier` is ok too if you want. Technically the schema is part of the table identifier. This will be clearer once we move to TypeScript

>  but even these strings could have schema qualifiers

They can't, if you pass `public.user`, it will be treated as `"public"."public.user"`. The only way to pass a schema is by doing `{ tableName: 'users', schema: 'users' }`. Passing `users` is strictly equivalent to passing `{ tableName: 'users' }`